### PR TITLE
Add Query/BySigner to x/data

### LINKED ISSUE #168: Query signatures by signer in x/data
It would be nice to be able to query all the pieces of data that an account is signed.

To do this we should likely store signatures as separate keys formed by `concat(cid, signer)` with a reverse index key `concat(signer, cid)`.

### PR DESCRIPTION
**Pull request by aaronc**
Closes #168

### CODE REVIEW

File: x/data/server/keys.go [hunk 1]
```diff
@@ -21,19 +20,29 @@ func CIDBase64String(cid []byte) string {
...
+	key := SignerCIDIndexPrefix(signer)
+	key = append(key, cid...)
+	return key
 }
 
 func SignerCIDIndexPrefix(signer string) []byte {
-	return append([]byte{SignerCIDPrefix}, []byte(fmt.Sprintf("%s/", signer))...)
+	key := []byte{SignerCIDPrefix}
+	key = append(key, signer...)
+	key = append(key, 0)
```
Comment (clevinson) [inline x/data/server/keys.go:44]: Any particular reason for switching to `0` instead of `/` as a delimeter? If we expect users to ever be directly querying KVStore via tendermint RPC, then it does probably make sense for us to think of this as an API surface.

If we do switch here, should we also change other prefix deliminters (e.g. ecocredit's batchDenom prefixes also use `/` [here](https://github.com/regen-network/regen-ledger/blob/master/x/ecocredit/server/msg_server.go#L65))?

File: x/data/server/msg_server.go [hunk 2]
```diff
@@ -100,7 +100,7 @@ func (s serverImpl) SignData(goCtx context.Context, request *data.MsgSignDataReq
 
 		store.Set(key, emptyBz)
 		// set reverse lookup key
-		store.Set(SignerCIDKey(signer, cidStr), emptyBz)
+		store.Set(SignerCIDKey(signer, cidBz), emptyBz)
```
Comment (clevinson) [inline x/data/server/msg_server.go:103]: Why is this changed here? Shouldn't we choose one way, and keep it consistent across `SignerCIDKey()` and `CIDSignerKey()`? Currently `CIDSignerKey()` still takes `cidStr string`...

**Review by clevinson — COMMENTED** [review]
Few small qs.. otherwise looks good.

File: x/data/server/keys.go (hunk 1 above)
Comment (aaronc) [reply x/data/server/keys.go:44]: Since both are strings, `0` is guaranteed to not be a string as it's the null terminator. `'/'` is a valid ASCII character

File: x/data/server/msg_server.go (hunk 2 above)
Comment (aaronc) [reply x/data/server/msg_server.go:103]: A string is not needed since this is the second part of the index key. It's an optimization. Maybe better to be consistent, but also good to optimize where possible no?

File: x/data/server/keys.go (hunk 1 above)
Comment (aaronc) [reply x/data/server/keys.go:44]: So this is a more generic way to separate two strings in an index key

File: third_party/proto/cosmos/base/query/v1beta1/pagination.proto [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+message PageRequest {
+  // key is a value returned in PageResponse.next_key to begin
+  // querying the next page most efficiently. Only one of offset or key
+  // should be set.
+  bytes key = 1;
+
+  // offset is a numeric offset that can be used when key is unavailable.
+  // It is less efficient than using key. Only one of offset or key should
+  // be set.
+  uint64 offset = 2;
```
Comment (robert-zaremba) [inline third_party/proto/cosmos/base/query/v1beta1/pagination.proto:22]: Is it possible to get just the next key instead of using `offset`?

File: x/data/server/keys.go (hunk 1 above)
Comment (robert-zaremba) [reply x/data/server/keys.go:44]: I saw somewhere a note to use `0` (probably in an ADR). It will be worth to add a documentation to this function and pointing to a specification if available.

File: x/data/server/msg_server.go (hunk 2 above)
Comment (robert-zaremba) [reply x/data/server/msg_server.go:103]: Yes!

File: x/data/server/testsuite/suite.go [hunk 4]
```diff
@@ -92,6 +92,14 @@ func (s *IntegrationTestSuite) TestScenario() {
 	s.Require().Equal([]string{s.addr1.String()}, queryRes.Signers)
 	s.Require().Empty(queryRes.Content)
 
+	// query data by signer
+	bySignerRes, err := s.queryClient.BySigner(s.ctx, &data.QueryBySignerRequest{
+		Signer: s.addr1.String(),
+	})
+	s.Require().NoError(err)
+	s.Require().NotNil(bySignerRes)
+	s.Require().Contains(bySignerRes.Cids, cidBz)
```
Comment (robert-zaremba) [inline x/data/server/testsuite/suite.go:101]: I wonder, why do we want to use the long calls, rather than declaring a variable at the top of the function:
```
require := s.Require()
```

File: third_party/proto/cosmos/base/query/v1beta1/pagination.proto (hunk 3 above)
Comment (robert-zaremba) [reply third_party/proto/cosmos/base/query/v1beta1/pagination.proto:22]: Also, why don't we use `oneof` here?

File: proto/regen/data/v1alpha1/query.proto [hunk 5]
```diff
@@ -32,4 +36,25 @@ message QueryByCidResponse {
...
+
+// QueryBySignerRequest is the Query/BySigner request type.
+message QueryBySignerRequest {
+
+    // signer is the address of the signer to query by.
+    string signer = 1;
+
+    // pagination is the PageRequest to use for pagination.
+    cosmos.base.query.v1beta1.PageRequest pagination = 2;
+}
```
Comment (robert-zaremba) [inline proto/regen/data/v1alpha1/query.proto:50]: NOTE: It's good to have it for verification, but in most of the production use-cases we will probably like to limit the blockchain load, and use off-chain DB for queries and analytics.

File: third_party/proto/cosmos/base/query/v1beta1/pagination.proto (hunk 3 above)
Comment (aaronc) [reply third_party/proto/cosmos/base/query/v1beta1/pagination.proto:22]: This was how the cosmos SDK pagination got implemented. My design I admit, but open to changing it

File: x/data/server/testsuite/suite.go (hunk 4 above)
Comment (aaronc) [reply x/data/server/testsuite/suite.go:101]: 🤷🏻‍♂️

File: third_party/proto/cosmos/base/query/v1beta1/pagination.proto (hunk 3 above)
Comment (aaronc) [reply third_party/proto/cosmos/base/query/v1beta1/pagination.proto:22]: https://github.com/cosmos/cosmos-sdk/issues/6191

**Review by clevinson — APPROVED** [review]
🎉
